File: ios/FullscreenHandlerExpoModule.swift
import BitmovinPlayer
import ExpoModulesCore

/**
 * Expo module for FullscreenHandler management with bidirectional communication.
 * Handles synchronous fullscreen state changes between native code and JavaScript.
 */
public class FullscreenHandlerExpoModule: Module {
    /// In-memory mapping from `nativeId`s to `FullscreenHandlerBridge` instances.
    private var fullscreenHandlers: Registry<FullscreenHandlerBridge> = [:]

    /// Dispatch group used for blocking thread while waiting for state change
    private let fullscreenChangeDispatchGroup = DispatchGroup()

    public func definition() -> ModuleDefinition {
        Name("FullscreenHandlerExpoModule")

        Events("onEnterFullscreen", "onExitFullscreen")

        AsyncFunction("registerHandler") { (nativeId: String) in
            DispatchQueue.main.async { [weak self] in
                guard let self, self.fullscreenHandlers[nativeId] == nil else {
                    return
                }
                self.fullscreenHandlers[nativeId] = FullscreenHandlerBridge(
                    nativeId,
                    moduleRegistry: appContext?.moduleRegistry
                )
            }
        }.runOnQueue(.main)

        AsyncFunction("destroy") { [weak self] (nativeId: String) in
            self?.fullscreenHandlers.removeValue(forKey: nativeId)
        }.runOnQueue(.main)

        AsyncFunction("notifyFullscreenChanged") { (nativeId: String, isFullscreenEnabled: Bool) -> Any? in
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                self.fullscreenHandlers[nativeId]?.isFullscreen = isFullscreenEnabled
                self.fullscreenChangeDispatchGroup.leave()
            }
            return nil
        }

        AsyncFunction("setIsFullscreenActive") { (nativeId: String, isFullscreen: Bool, promise: Promise) in
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                self.fullscreenHandlers[nativeId]?.isFullscreen = isFullscreen
                promise.resolve()
            }
        }.runOnQueue(.main)
    }

    /**
     * Retrieves the FullscreenHandlerBridge instance for the given nativeId.
     * This method maintains the same static access pattern as the legacy module.
     */
    @objc
    func retrieve(_ nativeId: String) -> FullscreenHandlerBridge? {
        fullscreenHandlers[nativeId]
    }

    /**
     * Handles fullscreen enter request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be entered.
     */
    func onFullscreenRequested(nativeId: String) {
        fullscreenChangeDispatchGroup.enter()

        // Send event to JavaScript
        sendEvent("onEnterFullscreen", [
            "nativeId": nativeId
        ])

//        fullscreenChangeDispatchGroup.wait()
    }

    /**
     * Handles fullscreen exit request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be exited.
     */
    func onFullscreenExitRequested(nativeId: String) {
        fullscreenChangeDispatchGroup.enter()

        // Send event to JavaScript
        sendEvent("onExitFullscreen", [
            "nativeId": nativeId
        ])

//        fullscreenChangeDispatchGroup.wait()
    }
}

File: android/src/main/java/com/bitmovin/player/reactnative/FullscreenHandlerExpoModule.kt
package com.bitmovin.player.reactnative

import com.bitmovin.player.reactnative.ui.FullscreenHandlerBridge
import com.facebook.react.bridge.*
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private const val MODULE_NAME = "FullscreenHandlerExpoModule"

/**
 * Expo module for FullscreenHandler management with bidirectional communication.
 * Handles synchronous fullscreen state changes between native code and JavaScript.
 */
class FullscreenHandlerExpoModule : Module() {
    /**
     * In-memory mapping from `nativeId`s to `FullscreenHandlerBridge` instances.
     */
    private val fullscreenHandlers: Registry<FullscreenHandlerBridge> = mutableMapOf()

    /**
     * Module's local lock object used to sync calls between Kotlin and JS.
     */
    private val lock = ReentrantLock()

    /**
     * Lock condition used to sync operations on the fullscreen handler.
     */
    private val fullscreenChangedCondition = lock.newCondition()

    override fun definition() = ModuleDefinition {
        Name(MODULE_NAME)

        AsyncFunction("registerHandler") { nativeId: String ->
            val fullscreenHandler = fullscreenHandlers[nativeId] ?: FullscreenHandlerBridge(appContext.reactContext as com.facebook.react.bridge.ReactApplicationContext, nativeId, this@FullscreenHandlerExpoModule)
            fullscreenHandlers[nativeId] = fullscreenHandler
        }

        AsyncFunction("destroy") { nativeId: String ->
            fullscreenHandlers.remove(nativeId)
        }

        AsyncFunction("notifyFullscreenChanged") { nativeId: String, isFullscreenEnabled: Boolean ->
            fullscreenHandlers[nativeId]?.isFullscreen = isFullscreenEnabled
            lock.withLock {
                fullscreenChangedCondition.signal()
            }
        }

        AsyncFunction("setIsFullscreenActive") { nativeId: String, isFullscreenActive: Boolean ->
            fullscreenHandlers[nativeId]?.isFullscreen = isFullscreenActive
        }
    }

    /**
     * Retrieves the FullscreenHandlerBridge instance for the given nativeId.
     * This method maintains the same static access pattern as the legacy module.
     */
    fun getInstance(nativeId: String?): FullscreenHandlerBridge? = fullscreenHandlers[nativeId]

    /**
     * Handles fullscreen enter request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be entered.
     */
    fun requestEnterFullscreen(nativeId: String) {
        // Call JavaScript function directly using React Native bridge
        appContext.reactContext?.let { context ->
            (context as ReactApplicationContext).catalystInstance.callFunction(
                "FullscreenBridge-$nativeId",
                "enterFullscreen",
                Arguments.createArray() as NativeArray
            )
        }
        
        lock.withLock {
            fullscreenChangedCondition.await()
        }
    }

    /**
     * Handles fullscreen exit request from native code.
     * Called by FullscreenHandlerBridge when fullscreen should be exited.
     */
    fun requestExitFullscreen(nativeId: String) {
        // Call JavaScript function directly using React Native bridge
        appContext.reactContext?.let { context ->
            (context as ReactApplicationContext).catalystInstance.callFunction(
                "FullscreenBridge-$nativeId",
                "exitFullscreen",
                Arguments.createArray() as NativeArray
            )
        }
        
        lock.withLock {
            fullscreenChangedCondition.await()
        }
    }

    companion object {
        /**
         * Static access method to maintain compatibility with other modules.
         * Retrieves the FullscreenHandlerBridge for the given nativeId.
         */
        @JvmStatic
        fun getFullscreenHandlerBridge(nativeId: String): FullscreenHandlerBridge? {
            // TODO: Implement global registry pattern if needed by other modules
            return null
        }
    }
}